DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/zero_copy_data_transfer.cpp

```diff
@@ -0,0 +1,184 @@
+    
+    // ensure that we could allocate space for both the input and output
+    if (in_restricted_usm == NULL) {
+      std::cerr << "ERROR: failed to allocate space for 'in_restricted_usm'\n";
+      return 1;
+    }
+    if (out_restricted_usm == NULL) {
+      std::cerr << "ERROR: failed to allocate space for 'out_restricted_usm'\n";
+      return 1;
+    }
+
+    // generate some random input data and compute the golden result
+    for (int i = 0; i < size; i++) {
+      // generate a random value
+      Type n = Type(rand() % 100);
+
+      // populate the inputs
+      in_buffer[i] = in_restricted_usm[i] = n;
+
+      // compute the golden result
+      out_gold[i] = n * i;
+    }
+
+    // run the buffer version kernel
+    std::cout << "Running the buffer kernel version with size=" << size << "\n";
+    std::vector<double> buffer_kernel_latency(iterations);
+    for (size_t i = 0; i < iterations; i++) {
+      buffer_kernel_latency[i] = BufferKernel<Type>(q, in_buffer,
+                                                    out_buffer, size);
+    }
+
+    // run the restricted USM version kernel
+    std::cout << "Running the restricted USM kernel version with size=" << size
+              << "\n";
+    std::vector<double> restricted_usm_latency(iterations);
+    for (size_t i = 0; i < iterations; i++) {
+      restricted_usm_latency[i] = RestrictedUSMKernel<Type>(q,
+                                                            in_restricted_usm,
+                                                            out_restricted_usm,
+                                                            size);
+    }
+
+    // validate the outputs
+    // validate the buffer version
+    for (int i = 0; i < size; i++) {
+      if (out_gold[i] != out_buffer[i]) {
+        std::cerr << "ERROR: mismatch at entry " << i
+                  << " of 'Buffer' kernel output "
+                  << "(" << out_gold[i] << "," << out_buffer[i] << ")"
+                  << "\n";
+        return 1;
+      }
+    }
+    // validate the restricted USM version
+    for (int i = 0; i < size; i++) {
+      if (out_gold[i] != out_restricted_usm[i]) {
+        std::cerr << "ERROR: mismatch at entry " << i
+                  << " of 'RestrictedUSM' kernel output "
+                  << "(" << out_gold[i] << "," << out_restricted_usm[i] << ")"
+                  << "\n";
+        return 1;
+      }
+    }
+
+    // The FPGA emulator does not accurately represent the hardware performance
+    // so we don't print performance results when running with the emulator
+#ifndef FPGA_EMULATOR
+    // Compute the average latency across all iterations.
+    // We use the first iteration as a 'warmup' for the FPGA,
+    // so we ignore its results.
+    double buffer_avg_lat = std::accumulate(buffer_kernel_latency.begin() + 1,
+                                            buffer_kernel_latency.end(), 0.0) /
+                            (double)(iterations - 1);
```

Comment:
```suggestion
                            (iterations - 1);
```
already a double because of 0.0